cdap-spark-core/src/main/scala/co/cask/cdap/app/runtime/spark/SparkContextCache.scala

```diff
@@ -131,17 +135,129 @@ object SparkContextCache {
         try {
           streamingContext.foreach(ssc => {
             ssc.stop(false)
-            ssc.awaitTermination()
+            ssc.awaitTermination
           })
         } finally {
-          sparkContext.foreach(_.stop())
+          sparkContext.foreach(sc => {
+            val cleanup = createCleanup(sc);
+            try {
+              sc.stop
+            } finally {
+              cleanup()
+            }
+          })
         }
       }
+      sparkListeners.clear()
       sparkContext;
     }
   }
 
   /**
+    * Creates a function that will stop and cleanup up all http servers and thread pools
+    * associated with the given SparkContext.
+    */
+  private def createCleanup(sc: SparkContext): () => Unit = {
+    val closers = ArrayBuffer.empty[Option[() => Unit]]
+
+    // Create a closer function for the file server in Spark
+    // The SparkEnv is either accessed through the "env() method (1.4+) or the "env" field in the SparkContext
+    // Every field access or method call are done through Option, hence if there is any changes in Spark,
+    // it won't fail the execution, although it might create permgen leakage, but it also depends on Spark
+    // has fixed this Thread resource leakage or not.
+    closers += sc.callMethod("env").orElse(sc.getField("env")).flatMap(env =>
+      env.getField("rpcEnv").flatMap(rpcEnv =>
+        rpcEnv.callMethod("fileServer").flatMap(_ match {
+          case s if "org.apache.spark.rpc.netty.HttpBasedFileServer" == s.getClass.getName =>
+            s.getField("httpFileServer")
+          case s if "org.apache.spark.HttpFileServer" == s.getClass.getName => Some(s)
```

Comment:
Actually this is a bug. After looking at the Spark source code again, the `fileServer` method returns one of the following three things:

1. a `HttpBaseFileServer`. This is returned by `NettyRpcEnv`, which is the default in Spark 1.6.
2. a `NettyStreamManager`. This is returned by `NettyRpcEnv` when `spark.rpc.useNettyFileServer` is set to `true` and also in Spark 1.7+
3. a `AkkaFileServer`. This is returned by `AkkaRpcEnv` when `spark.rpc` is set to `akka`. In this case, the `HttpFileServer` that we need to access is in the `httpFileServer` field.

For Spark version <= 1.5, there is only akka and they use akka as the file transport.